sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -221,6 +221,106 @@ object DataSourceAnalysis extends Rule[LogicalPlan] with CastSupport {
+ * 2. Repartition by dynamic partition column when writing dynamic partition table/directory.
+ * 3. Repartition by bucket column with bucket number and sort by sort column when writing
+ *    bucket table/directory.
+ *
+ * Note that, this rule must be run after `DataSourceAnalysis`.
+ */
+object RepartitionWritingDataSource extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (conf.repartitionWritingDataSource) {
+      insertRepartition(plan)
+    } else {
+      plan
+    }
+  }
+
+  private def insertRepartition(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case c @ CreateDataSourceTableAsSelectCommand(table, _, query, _) if c.resolved =>
+      val dynamicPartExps = resolveColumnNames(table.partitionColumnNames, query.output)
+      applyRepartition(c, table.bucketSpec, dynamicPartExps)
+
+    case i: InsertIntoHadoopFsRelationCommand if i.resolved =>
+      val dynamicPartExps = i.partitionColumns
+        .filterNot(p => i.staticPartitions.exists(s => conf.resolver(p.name, s._1)))
+      applyRepartition(i, i.bucketSpec, dynamicPartExps)
+
+    case i: InsertIntoDataSourceDirCommand if i.resolved && canApplyRepartition(i.query) =>
+      i.copy(query = RepartitionByExpression(Nil, i.query, None))
+
+    // InsertIntoDataSourceCommand only accept InsertableRelation, do not need repartition.
+  }
+
+  def applyRepartition(
+      dataWritingCommand: DataWritingCommand,
+      bucketSpec: Option[BucketSpec],
+      partitionColumns: Seq[Expression]): LogicalPlan = {
+    val query = dataWritingCommand.query
+    (bucketSpec, partitionColumns) match {
+      case (None, Nil) =>
+        if (canApplyRepartition(query)) {
+          dataWritingCommand.withNewChildrenInternal(
+            IndexedSeq(RepartitionByExpression(Nil, query, None)))
+        } else {
+          dataWritingCommand
+        }
+
+      case (None, partExps @ _ +: _) =>
+        query match {
+          case RepartitionByExpression(partExpressions, _, _) if partExpressions == partExps =>
+            dataWritingCommand
+          case _ =>
+            dataWritingCommand.withNewChildrenInternal(
+              IndexedSeq(RepartitionByExpression(partExps, query, None)))
+        }
+
+      case (Some(bucket), _) if bucket.sortColumnNames.nonEmpty =>
+        val bucketExps = resolveColumnNames(bucket.bucketColumnNames, query.output)
+        val sortExps = resolveColumnNames(bucket.sortColumnNames, query.output)
+          .map(SortOrder(_, Ascending))
+        query match {
+          case Sort(order, false, RepartitionByExpression(partExpr, _, Some(bucket.numBuckets)))
+            if order == sortExps && partExpr == bucketExps =>
+            dataWritingCommand
+          case _ =>
+            dataWritingCommand.withNewChildrenInternal(
+              IndexedSeq(Sort(sortExps, false,
+                RepartitionByExpression(bucketExps, query, Some(bucket.numBuckets)))))
```

Comment:
Why we don't repartition by partitionColumns for bucketed table if it's a dynamic partition query ?